db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Some parts of this source code are based on Apache Derby, and the following notices apply to
+ * Apache Derby:
+ *
+ * Apache Derby is a subproject of the Apache DB project, and is licensed under
+ * the Apache License, Version 2.0 (the "License"); you may not use these files
+ * except in compliance with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Splice Machine, Inc. has modified the Apache Derby code in this file.
+ *
+ * All such Splice Machine modifications are Copyright 2012 - 2020 Splice Machine, Inc.,
+ * and are licensed to you under the GNU Affero General Public License.
+ */
+
+package com.splicemachine.db.impl.jdbc.authentication;
+
+import com.splicemachine.db.authentication.AccessTokenVerifier;
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.reference.Attribute;
+import com.splicemachine.db.iapi.services.monitor.Monitor;
+import com.splicemachine.db.iapi.sql.dictionary.DataDictionary;
+import com.splicemachine.db.iapi.sql.dictionary.UserDescriptor;
+import com.splicemachine.db.iapi.util.IdUtil;
+import com.splicemachine.db.impl.jdbc.Util;
+
+import java.lang.reflect.Method;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * <p>
+ * This authentication service supports Derby NATIVE and TOKEN authentication.
+ * </p>
+ *
+ * <p>
+ * To activate this service, set the db.authentication.provider database
+ * or system property to a value beginning with the token "TOKEN:".
+ * </p>
+ *
+ * <p>
+ * This service instantiates and calls the basic User authentication scheme at runtime.
+ * </p>
+ *
+ * <p>
+ * User credentials are defined in the SYSUSERS table.
+ * </p>
+ */
+public final class TokenAuthenticationServiceImpl extends NativeAuthenticationServiceImpl {
+
+    public String authenticateUser (String userName, String userPassword, String databaseName, Properties info) throws SQLException {
```

Comment:
I think we need a better mechanism to compose multiple authentication services at the same time. Can you open a Jira that removes the inheritance here and addresses the composability?

Reply:
https://splicemachine.atlassian.net/browse/DB-11554